### PR TITLE
[GH-1305] Refactor test_dtypes.py to be cleaner

### PR DESCRIPTION
**Pull request by waifairer**
Pretty straightforward DRY refactor of the test_dtypes.py file.

### CODE REVIEW

**Comment by waifairer** [comment]
Fairly close to asking for review here

**Comment by waifairer** [comment]
Apologies for the sort of stiff, corporate issue tracking here. It'll take me a bit to loosen up

File: test/test_dtype.py [hunk 1]
```diff
@@ -3,28 +3,70 @@
...
+  if target_contents is None: target_contents = copy.deepcopy(tensor_contents)
+  list(map(
+    lambda t_dtype: _test_cast(Tensor(copy.deepcopy(tensor_contents), dtype=source_dtype), t_dtype, copy.deepcopy(tensor_contents)),
+    target_dtypes
+  ))
+# tests no-op casts from source_dtypes to target_dtype
+def _test_casts_to(tensor_contents:List, source_dtypes:List[DType], target_dtype:DType, target_contents:Optional[List]=None):
+  if target_contents is None: target_contents = copy.deepcopy(tensor_contents)
+  list(map(
+    lambda s_dtype: _test_cast(Tensor(copy.deepcopy(tensor_contents), dtype=s_dtype), target_dtype, target_contents),
```
Comment (wozeparrot) [inline test/test_dtype.py:51]: what's the need of the `deepcopy` here and above?

File: test/test_dtype.py [hunk 2]
```diff
@@ -3,28 +3,70 @@
...
+  _assert_eq(c, target_dtype, target)
+
+def _test_cast(a:Tensor, target_dtype:DType, target):
+  _test_op(lambda: a.cast(target_dtype), target_dtype, target)
+# tests no-op casts from source_dtype to target_dtypes
+def _test_casts_from(tensor_contents:List, source_dtype:DType, target_dtypes:List[DType], target_contents:Optional[List]=None
+                     ):
+  if target_contents is None: target_contents = copy.deepcopy(tensor_contents)
+  list(map(
+    lambda t_dtype: _test_cast(Tensor(copy.deepcopy(tensor_contents), dtype=source_dtype), t_dtype, copy.deepcopy(tensor_contents)),
```
Comment (waifairer) [inline test/test_dtype.py:44]: I suppose I wouldn't need to deepcopy them both.

File: test/test_dtype.py [hunk 3]
```diff
@@ -3,28 +3,70 @@
...
+  if target_contents is None: target_contents = copy.deepcopy(tensor_contents)
+  list(map(
+    lambda s_dtype: _test_cast(Tensor(tensor_contents, dtype=s_dtype), target_dtype, target_contents),
+    source_dtypes
+  ))
+
+def _test_ops(a_dtype:DType, b_dtype:DType, target_dtype:DType):
+  _assert_eq(
+    Tensor([1,2,3,4], dtype=a_dtype) + Tensor([1,2,3,4], dtype=b_dtype),
+
```
Comment (waifairer) [inline test/test_dtype.py:58]: the linter must not be very sophisticated cause it didn't catch this

File: test/test_dtype.py (hunk 1 above)
Comment (waifairer) [reply test/test_dtype.py]: Yeah i was looking at that too. Pushed, only one deepcopy now

File: test/test_dtype.py (hunk 3 above)
Comment (waifairer) [reply test/test_dtype.py]: maybe I'm being hard on it

File: test/test_dtype.py [hunk 4]
```diff
@@ -3,28 +3,69 @@
 from tinygrad.helpers import getenv, DType, DEBUG
 from tinygrad.lazy import Device
 from tinygrad.tensor import Tensor, dtypes
+from typing import List, Optional
 from extra.utils import OSX, temp
+import copy
+
```
Comment (wozeparrot) [inline test/test_dtype.py:10]: extra newline here

File: test/test_dtype.py [hunk 5]
```diff
@@ -3,28 +3,69 @@
...
+    raise AssertionError(
+      f"\ntensor {tensor.numpy()} dtype {tensor.dtype} does not match target {target} with dtype {target_dtype}"
+    ) from e
 
 def _test_op(fxn, target_dtype:DType, target):
   c = fxn()
-  if DEBUG >= 2: print(c.numpy())
-  assert c.dtype == target_dtype
-  np.testing.assert_allclose(c.numpy(), target)
+  _assert_eq(c, target_dtype, target)
```
Comment (wozeparrot) [inline test/test_dtype.py:35]: Can just `_assert_eq(fxn(), target_dtype, target)`

File: test/test_dtype.py [hunk 6]
```diff
@@ -3,28 +3,69 @@
...
+
+def _test_cast(a:Tensor, target_dtype:DType, target):
+  _test_op(lambda: a.cast(target_dtype), target_dtype, target)
+# tests no-op casts from source_dtype to target_dtypes
+def _test_casts_from(tensor_contents:List, source_dtype:DType, target_dtypes:List[DType], target_contents:Optional[List]=None
+                     ):
```
Comment (wozeparrot) [inline test/test_dtype.py:41]: formatting

File: test/test_dtype.py [hunk 7]
```diff
@@ -3,28 +3,69 @@
...
+  if target_contents is None: target_contents = copy.deepcopy(tensor_contents)
+  list(map(
+    lambda t_dtype: _test_cast(Tensor(tensor_contents, dtype=source_dtype), t_dtype, target_contents),
+    target_dtypes
+  ))
+# tests no-op casts from source_dtypes to target_dtype
+def _test_casts_to(tensor_contents:List, source_dtypes:List[DType], target_dtype:DType, target_contents:Optional[List]=None):
+  if target_contents is None: target_contents = copy.deepcopy(tensor_contents)
```
Comment (wozeparrot) [inline test/test_dtype.py:49]: I still don't understand why there needs to be a deepcopy here. Tensor *should* make a copy from a python list internally i think.

File: test/test_dtype.py [hunk 8]
```diff
@@ -3,28 +3,69 @@
 from tinygrad.helpers import getenv, DType, DEBUG
 from tinygrad.lazy import Device
 from tinygrad.tensor import Tensor, dtypes
+from typing import List, Optional
 from extra.utils import OSX, temp
+import copy
+
 
 def _test_to_np(a:Tensor, np_dtype, target):
   print(a)
```
Comment (wozeparrot) [inline test/test_dtype.py:12]: This should probably be gated behind `DEBUG=2` as well, it wasn't before but it makes CI a *tiny tiny tiny* bit faster.

**Review by wozeparrot — CHANGES_REQUESTED** [review]

File: test/test_dtype.py (hunk 7 above)
Comment (waifairer) [reply test/test_dtype.py:49]: Oh, I think I'm just cautious but I can follow Tensor in and see.

File: test/test_dtype.py (hunk 5 above)
Comment (waifairer) [reply test/test_dtype.py:35]: Done.

File: test/test_dtype.py (hunk 7 above)
Comment (waifairer) [reply test/test_dtype.py:49]: I don't think numpy.array copies it. That's all tensor does with it -- numpy.array(data)

Comment (waifairer) [reply test/test_dtype.py:49]: I think I'd rather keep the deep copy here to be safe.

File: test/test_dtype.py (hunk 8 above)
Comment (waifairer) [reply test/test_dtype.py:12]: Done

File: test/test_dtype.py [hunk 9]
```diff
@@ -3,28 +3,66 @@
...
+    target_dtype, [2,4,6,8]
+  )
+  _assert_eq(
+    Tensor([1,2,3,4], dtype=a_dtype) * Tensor([1,2,3,4], dtype=b_dtype),
+    target_dtype, [1,4,9,16]
+  )
+  _assert_eq(
+    Tensor([[1,2],[3,4]], dtype=a_dtype) @ Tensor.eye(2, dtype=b_dtype),
+    target_dtype, [[1,2],[3,4]]
+  )
```
Comment (wozeparrot) [inline test/test_dtype.py:64]: Can you just one line these?

Comment (waifairer) [reply test/test_dtype.py]: Yep, no problem. One-lined these and a few other things that were easy to read.

File: test/test_dtype.py [hunk 10]
```diff
@@ -3,28 +3,51 @@
...
+  _assert_eq(Tensor([1,2,3,4], dtype=a_dtype)*Tensor([1,2,3,4], dtype=b_dtype), target_dtype, [1,4,9,16])
+  _assert_eq(Tensor([[1,2],[3,4]], dtype=a_dtype)@Tensor.eye(2, dtype=b_dtype), target_dtype, [[1,2],[3,4]])
 
-def _test_cast(a:Tensor, target_dtype:DType, target): _test_op(lambda: a.cast(target_dtype), target_dtype, target)
-def _test_add(a:Tensor, b:Tensor, target_dtype:DType, target): _test_op(lambda: a+b, target_dtype, target)
-def _test_mul(a:Tensor, b:Tensor, target_dtype:DType, target): _test_op(lambda: a*b, target_dtype, target)
-def _test_matmul(a:Tensor, b:Tensor, target_dtype:DType, target): _test_op(lambda: a@b, target_dtype, target)
-def _test_add_upcast(a:Tensor, b:Tensor, target_dtype:DType, target): _test_op(lambda: a+b, target_dtype, target)
-def _test_mul_upcast(a:Tensor, b:Tensor, target_dtype:DType, target): _test_op(lambda: a*b, target_dtype, target)
-def _test_matmul_upcast(a:Tensor, b:Tensor, target_dtype:DType, target): _test_op(lambda: a@b, target_dtype, target)
```
Comment (wozeparrot) [inline test/test_dtype.py:52]: extra newlines

Comment (waifairer) [reply test/test_dtype.py:52]: Ok.

Comment (waifairer) [reply test/test_dtype.py:52]: Do you think we're ready for pylinter to take care of this? I would be happy to tune it.

Comment (waifairer) [reply test/test_dtype.py]: 1-2 newlines between classes seems ok, but 3 is definitely too many. I can make that a rule. Defs can have a maximum of 1 newline between them.

**Review by wozeparrot — APPROVED** [review]

**Comment by waifairer** [comment]
🎉 we're in sync!

**Pull request merged**